### PR TITLE
fix: loguru exception handling for invalid recordings

### PR DESCRIPTION
**Pull request by 0dm**
**What kind of change does this PR introduce?**

raises errors for invalid recordings

**Summary**

when attempting to visualize with a broken recording (not written properly if interrupted early), you may get a message like:
```
  File "/Users/aaron/Documents/GitHub/OpenAdapt/openadapt/app/tray.py", line 108, in _visualize
    visualize(recording)
  File "/Users/aaron/Documents/GitHub/OpenAdapt/openadapt/visualize.py", line 163, in main
    action_events = get_events(recording, process=PROCESS_EVENTS, meta=meta)
  File "/Users/aaron/Documents/GitHub/OpenAdapt/openadapt/events.py", line 22, in get_events
    screenshots = crud.get_screenshots(recording)
  File "/Users/aaron/Documents/GitHub/OpenAdapt/openadapt/crud.py", line 240, in get_screenshots
    screenshots[0].prev = screenshots[0]
IndexError: list index out of range
```
or 
```
  File "/Users/aaron/Documents/GitHub/OpenAdapt/openadapt/visualize.py", line 163, in main
    action_events = get_events(recording, process=PROCESS_EVENTS, meta=meta)
  File "/Users/aaron/Documents/GitHub/OpenAdapt/openadapt/events.py", line 46, in get_events
    action_events, window_events, screenshots = process_events(
  File "/Users/aaron/Documents/GitHub/OpenAdapt/openadapt/events.py", line 633, in process_events
    pct_action_events = num_action_events_ / num_action_events
ZeroDivisionError: division by zero
```

now, we get messages like:
```
2023-07-07 12:00:27.346 | ERROR    | openadapt.visualize:main:162 - An error has been caught in function 'main', process 'MainProcess' (5263), thread 'MainThread' (8153144128):
Traceback (most recent call last):

  File "<string>", line 1, in <module>

> File "/Users/aaron/Documents/GitHub/OpenAdapt/openadapt/visualize.py", line 162, in main
    action_events = get_events(recording, process=PROCESS_EVENTS, meta=meta)
                    |          |                  |                    -> {}
                    |          |                  -> True
                    |          -> Recording(id=1, timestamp=1688745592.017052, monitor_width=1440, monitor_height=900, double_click_interval_seconds=0.5, doubl...
                    -> <function get_events at 0x16a9f7c70>

  File "/Users/aaron/Documents/GitHub/OpenAdapt/openadapt/events.py", line 30, in get_events
    raise ValueError("No action events found.")

ValueError: No action events found.
```
or 
```
2023-07-07 11:58:53.627 | ERROR    | openadapt.events:get_events:20 - An error has been caught in function 'get_events', process 'MainProcess' (4998), thread 'MainThread' (8153144128):
Traceback (most recent call last):

  File "<string>", line 1, in <module>

  File "/Users/aaron/Documents/GitHub/OpenAdapt/openadapt/visualize.py", line 162, in main
    action_events = get_events(recording, process=PROCESS_EVENTS, meta=meta)
                    |          |                  |                    -> {}
                    |          |                  -> True
                    |          -> None
                    -> <function get_events at 0x13b6dbf40>

> File "/Users/aaron/Documents/GitHub/OpenAdapt/openadapt/events.py", line 20, in get_events
    action_events = crud.get_action_events(recording)
                    |    |                 -> None
                    |    -> <function get_action_events at 0x13b6dbc70>
                    -> <module 'openadapt.crud' from '/Users/aaron/Documents/GitHub/OpenAdapt/openadapt/crud.py'>

  File "/Users/aaron/Documents/GitHub/OpenAdapt/openadapt/crud.py", line 161, in get_action_events
    raise ValueError("Invalid recording")

ValueError: Invalid recording
```

**Checklist**
* [x] My code follows the style guidelines of OpenAdapt
* [x] I have performed a self-review of my code
* [x] If applicable, I have added tests to prove my fix is functional/effective
* [x] I have linted my code locally prior to submission
* [x] I have commented my code, particularly in hard-to-understand areas
* [x] I have made corresponding changes to the documentation (e.g. README.md, requirements.txt)
* [x] New and existing unit tests pass locally with my changes

**How can your code be run and tested?**

**Other information**

### CODE REVIEW

File: openadapt/crud.py [hunk 1]
```diff
@@ -155,7 +156,10 @@ def _get(table, recording_timestamp):
     )
 
 
+@logger.catch(reraise=True, onerror=lambda _: sys.exit(1))
 def get_action_events(recording):
+    if recording is None:
+        raise ValueError("Invalid recording.")
```
Comment (abrichr) [inline openadapt/crud.py:162]: How about just
```
assert recording, recording
```

File: openadapt/events.py [hunk 2]
```diff
@@ -25,6 +27,8 @@ def get_events(
     logger.debug(f"raw_action_event_dicts=\n{pformat(raw_action_event_dicts)}")
 
     num_action_events = len(action_events)
```
Comment (abrichr) [inline openadapt/events.py:29]: How about:
```
assert action_events, action_events
```

Comment (0dm) [reply openadapt/events.py:29]: I've done assert num_action_events > 0, "No action events found." instead, to produce a message.

File: openadapt/crud.py (hunk 1 above)
Comment (0dm) [reply openadapt/crud.py]: `assert recording, "Invalid recording."`

**Comment by 0dm** [comment]
@abrichr does this look good? You can reproduce these messages with the following steps:

### invalid recording
1. `poetry run record "test"`
2. send keyboard interrupt immediately
3. `poetry run visualize`
image


### no action events found
1. `poetry run record "test"`
2. click around the terminal for a second and stop the recording before the event writers start 
3.  `poetry run visualize`
image

File: openadapt/crud.py [hunk 3]
```diff
@@ -155,7 +156,9 @@ def _get(table, recording_timestamp):
     )
 
 
+@logger.catch(reraise=True, onerror=lambda _: sys.exit(1))
```
Comment (abrichr) [inline openadapt/crud.py:159]: Why do we need both attributes? Won't `sys.exit` break pytest?

File: openadapt/replay.py [hunk 4]
```diff
@@ -10,7 +10,7 @@
 
 LOG_LEVEL = "INFO"
 
-
+@logger.catch()
```
Comment (abrichr) [inline openadapt/replay.py:13]:
```suggestion
@logger.catch
```

File: openadapt/visualize.py [hunk 5]
```diff
@@ -149,7 +149,7 @@ def dict2html(obj, max_children=MAX_TABLE_CHILDREN, max_len=MAX_TABLE_STR_LEN):
             html_str = head + middle + tail
     return html_str
 
-
+@logger.catch()
```
Comment (abrichr) [inline openadapt/visualize.py:152]:
```suggestion
@logger.catch
```

File: openadapt/events.py [hunk 6]
```diff
@@ -1,3 +1,4 @@
+import sys
```
Comment (abrichr) [inline openadapt/events.py:1]:
```suggestion
```

File: openadapt/crud.py [hunk 7]
```diff
@@ -1,3 +1,4 @@
+import sys
```
Comment (abrichr) [inline openadapt/crud.py:1]:
```suggestion
```

**Pull request merged**